pkg/kubelet/kubelet.go

```diff
@@ -2737,6 +2747,139 @@ func (kl *Kubelet) fastStatusUpdateOnce() {
 	}, 100*time.Millisecond, stopCh)
 }
 
+// handleMaxContainerCheckpoints goes through the list of checkpoints of a
+// container in a pod and deletes old checkpoints if there are more than
+// 'kl.maxContainerCheckpoints'. To track the number of checkpoints a JSON file
+// is created and updated for each new checkpoint. If a new checkpoint would
+// result in more than 'kl.maxContainerCheckpoints' the oldest checkpoint will
+// be removed.
+func (kl *Kubelet) handleMaxContainerCheckpoints(
+	podUID types.UID,
+	podFullName,
+	containerName string,
+	options *runtimeapi.CheckpointContainerRequest,
+	checkpointTime int64,
+) error {
+
+	if kl.maxContainerCheckpoints <= 0 {
+		// This cannot really happen as the kubelet config parser will
+		// set this to 10 as a default value if 0 or not set. The error
+		// handling is mainly here to make sure tests are working.
+		return errors.New("maxContainerCheckpoints needs to be > 0")
+	}
+
+	checkpointJsonName := fmt.Sprintf(
+		"checkpoint-%s-%s.json",
+		podFullName,
+		containerName,
+	)
+
+	checkpointJsonPath := filepath.Join(kl.getCheckpointsDir(), checkpointJsonName)
+	// Lock while accessing the checkpoints on disk to avoid chaos
+	// if multiple threads would do it at the same time.
+	kl.updateMaxContainerCheckpoints.Lock()
+	defer kl.updateMaxContainerCheckpoints.Unlock()
+
+	checkpointMetadata, err := func() (
+		*checkpointctl.KubernetesContainerCheckpointMetadata,
+		error,
+	) {
+		_, err := os.Stat(checkpointJsonPath)
+		// If the file does not exist return an empty structure.
+		if err != nil {
+			return &checkpointctl.KubernetesContainerCheckpointMetadata{}, nil
+		}
+		ccm := &checkpointctl.KubernetesContainerCheckpointMetadata{}
+		_, err = checkpointctl.ReadJSONFile(
+			ccm,
+			kl.getCheckpointsDir(),
+			checkpointJsonName,
+		)
+		if err != nil {
+			return nil, fmt.Errorf(
+				"failed to read checkpoint metadata file (%s): %w",
+				checkpointJsonPath,
+				err,
+			)
+		}
+		return ccm, nil
+	}()
+
+	if err != nil {
+		return err
+	}
+
+	fi, err := os.Stat(options.Location)
+	if err != nil {
+		return fmt.Errorf(
+			"failed to stat checkpoint archive (%s): %w",
+			options.Location,
+			err,
+		)
+	}
+
+	numberOfCheckpoints := int32(len(checkpointMetadata.Checkpoints))
+
+	for numberOfCheckpoints >= kl.maxContainerCheckpoints {
+		klog.InfoS(
+			"More than maxContainerCheckpoints",
+			"container", containerName,
+			"numberOfCheckpoints", numberOfCheckpoints,
+			"maxContainerCheckpoints", kl.maxContainerCheckpoints,
+		)
+		// Find oldest checkpoint and delete it.
+		oldest := (checkpointMetadata.Checkpoints)[0].Timestamp
```

Comment:
We don't need those additional `(` `)` here, right?

Reply:
right, removed